### ISSUE TITLE
Marching cubes bug

### ISSUE THREAD

**Issue by lorinczszabolcs** [pyvista/pyvista#3387]
### Describe the bug, what's wrong, and what you expected.

I am trying to run the marching cubes algorithm on a 3D binary array, but I get the following error:
`Number of scalars (47448064) must match either the number of points (47448064) or the number of cells (47001780).`
Funny part is, that the number of scalars actually matches the number of points.

### Steps to reproduce the bug.

```python
dims=(181, 512, 512)
spacing=(2.5, 1.25, 1.25)
grid = pv.UniformGrid(
    dims=dims,
    spacing=spacing,
    origin=(0,0,0),
)
x, y, z = grid.points.T
mesh = grid.contour([0.5], np.random.randn(dims[0], dims[1], dims[2]), method='marching_cubes')
```

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Fri Sep 30 10:30:43 2022 EEST

                OS : Linux
            CPU(s) : 8
           Machine : x86_64
      Architecture : 64bit
               RAM : 11.7 GiB
       Environment : Jupyter
       File system : ext4
        GPU Vendor : Microsoft Corporation
      GPU Renderer : D3D12 (NVIDIA GeForce RTX 3060 Laptop GPU)
       GPU Version : 3.3 (Core Profile) Mesa 21.2.6

  Python 3.9.12 (main, Apr  5 2022, 06:56:58)  [GCC 7.5.0]

           pyvista : 0.36.1
               vtk : 9.1.0
             numpy : 1.22.4
           imageio : 2.9.0
           appdirs : 1.4.4
            scooby : 0.5.12
        matplotlib : 3.5.1
             PyQt5 : 5.15.7
           IPython : 8.2.0
          colorcet : 1.0.0
           cmocean : 2.0
        ipyvtklink : 0.2.2
             scipy : 1.7.3
              tqdm : 4.64.0
            meshio : 5.3.4
        jupyterlab : 3.3.2
         pythreejs : Version unknown

  Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904
  for Intel(R) 64 architecture applications
--------------------------------------------------------------------------------
```

### Screenshots

![image](https://user-images.githubusercontent.com/13637835/193215879-f78bbbfb-97a1-4d8b-b67b-1ce0d7fe6b1d.png)

**Comment by MatthewFlamm** [comment]
```py
mesh = grid.contour([0.5], np.ravel(np.random.randn(dims[0], dims[1], dims[2])), method='marching_cubes')
```

In pyvista, the first dimension has the match the number of points or cells.  So you need to flatten or ravel your array.  I haven't checked that it is flattened in the correct order in this example.

This comes up often enough that I think we should change the error message in case the total size of the array matches the number of points.

**Comment by lorinczszabolcs** [comment]
Thank you for the quick reply!

Ah, indeed. Yes, the error message could be more explicit.

Alternatively, could there be a check for the array size and try to infer the intended shape automatically?

**Comment by MatthewFlamm** [comment]
> Alternatively, could there be a check for the array size and try to infer the intended shape automatically?

This would be great, but it probably has some nasty edge cases that would need to be handled.  For example, we have 9 points and 9 cells.  The user supplies an array of size (3x3).  Was this cell data or point data, was it intended to be vector valued data, but the user should have supplied (9x3) shape data?  Unless this can be clarified in a straightforward way, I think having some more helpful error messages would be a good first step.

**Comment by banesullivan** [comment]
> This would be great, but it probably has some nasty edge cases that would need to be handled.

Lots of edge cases indeed

FWIW, `vtkImageData`/`pyvista.UniformGrid` can handle this sort of implicit matching pretty easily and I believe @akaszynski was working on this recently?

Otherwise, I think the best way to resolve this issue is to improve the error messages so that they are more explicit about the first dimension needing to match the number of points/cells and hinting towards calling `.ravel(order='F')`

**Comment by MatthewFlamm** [comment]
I'm going to put this as a good-first-issue. I _think_ we are suggesting that we could check for multidimensional data and then provide a more helpful error message in the case where `provided_array.size == mesh.n_points or provided_array.size == mesh.n_cells `` with some suggestions to consider `np.ravel` or `ndarray.flatten`.

**Comment by akaszynski** [comment]
> I believe @akaszynski was working on this recently?

Yep, that's part of the textures PR #2729. Really want that merged before next release.
